server/devpi_server/views.py

```diff
@@ -1446,6 +1446,104 @@ def user_list(self):
+        notices = []
+        tot_wheels = 0
+        tot_project_versions = 0
+        if context.user.getstage(target_index):
+            apireturn(409, "index %r already exists" % target_index)
+        try:
+            target_stage = context.user.create_stage(target_index)
+            for project_name in projects:
+                print(f"doing {project_name} ..")
+                project_versions = source_stage.list_versions(project_name)
+                tot_project_versions += len(project_versions)
+                for version in project_versions:
+                    print(f" {project_name}-{version} ..")
+                    try:
+                        linkstore = source_stage.get_linkstore_perstage(project_name, version)
+                        release_links = linkstore.get_links("releasefile", None)
+                        verdata = source_stage.get_versiondata_perstage(project_name, version)
+                    except Exception as err:
+                        failure = f"failed during read of {project_name}-{version}: {err}\n{traceback.format_exc()}"
+                        print(failure)
+                        failures.append(failure)
+                        continue
+                    # fixup verdata coming from source_stage for target_stage:
+                    new_elinks = []
+                    elinks = verdata.get('+elinks', None)
+                    if not elinks:
+                        msg = f"{project_name}-{version} has no +elinks: {elinks}"
+                        notices.append(msg)
+                        print(msg)
+                        continue
+                    for elink in elinks:
+                        try:
+                            front = source_stage.name + "/+f/"
+                            entrypath = elink['entrypath']
+                            assert entrypath.startswith(front), elink
+                            new_elink = dict(elink)
+                            new_elink['entrypath'] = entrypath.replace(front, f"{context.user.name}/{target_index}/+f/")
+                            new_elinks.append(new_elink)
+                        except Exception as err:
+                            msg = f"error with elink {elink}: {err}"
+                            failures.append(msg)
+                            print(msg)
+                    # have to set on _data cause `verdata` itself is ReadOnly:
+                    verdata._data['+elinks'] = new_elinks
+                    # fixup done.
+                    # we now have to set that version data into the target stage:
+                    target_stage.set_versiondata(verdata)
+                    # then we can create the (hard-)links for each release:
+                    for link in release_links:
+                        print(f"  {link.entrypath}")
+                        new_link = target_stage.store_releasefile(
+                            project_name, version,
+                            link.basename, link)
```

Comment:
I previously wrote to create a new link. No need for that. Use ``with`` on ``file_open_read``, set ``devpi_srcpath`` when a snapshot instead of a copy is requested and pass that in. That should also keep resource usage low and the existing code for all cases is used.

Reply:
ok. read all your comments. I'm not sure I get everything with `file_open_read` .. 

but I can see to create a new ELink in this main loop and pass that instead of the original link here .. and add a devpi_srcpath attribute to it (it does not have right now).

looking all that.. 

but FWIW: I had to `target_stage.set_versiondata(verdata)` before .. otherwise there was missing registration error.

Reply:
Instead of creating ``new_release_link`` and passing that in do something like this:

```
with link.entry.file_open_read() as f:
    f.devpi_srcpath = ...
    new_link = linkstore.create_linked_entry(
        rel="releasefile",
        basename=link.basename,
        content_or_file=f,
        last_modified=None)
    ...
```

Reply:
got it. made it like that. even better effectively. had also only to add the hash_spec and pass it. so to not have to re-compute it effectively.